File: test/python/test_keystore.py
import unittest
import json
import os
import sys
import asyncio
import tempfile
import shutil
from pathlib import Path
from datetime import datetime, timedelta

# Add project root to Python path to ensure imports work
project_root = Path(__file__).parents[2]
sys.path.append(str(project_root))

# Import the Keystore module
try:
    from hallucinate_app.python.hallucinate_app.keystore import Keystore
except ImportError:
    # Try alternative import path
    try:
        from hallucinate_app.hallucinate_app.keystore import Keystore
    except ImportError:
        print("Failed to import Keystore module. Make sure the module exists and paths are correct.")
        sys.exit(1)

class TestKeystore(unittest.TestCase):
    """Test the Keystore module for secure credential storage"""
    
    @classmethod
    def setUpClass(cls):
        # Create a temporary directory for testing
        cls.test_dir = tempfile.mkdtemp(prefix="hallucinate_app_keystore_test_")
        
        # Test encryption key
        cls.test_key = "test_master_key_do_not_use_in_production"
        
        # Initialize the keystore
        cls.keystore = Keystore(metadata={
            "encryption_key": cls.test_key,
            "storage_location": cls.test_dir
        })
        
        # Initialize event loop for async tests
        cls.loop = asyncio.get_event_loop()
        if cls.loop.is_closed():
            cls.loop = asyncio.new_event_loop()
            asyncio.set_event_loop(cls.loop)
        
        # Initialize the keystore
        cls.loop.run_until_complete(cls.keystore.init())
    
    @classmethod
    def tearDownClass(cls):
        # Remove the temporary directory
        shutil.rmtree(cls.test_dir)
    
    def test_initialization(self):
        """Test that the keystore initialized correctly"""
        self.assertTrue(self.keystore.initialized)
        self.assertTrue(os.path.exists(self.test_dir))
    
    def test_key_storage_and_retrieval(self):
        """Test storing and retrieving keys"""
        test_provider = "test_provider"
        test_key = f"test_api_key_{int(datetime.now().timestamp())}"
        
        # Store the key
        result = self.loop.run_until_complete(
            self.keystore.set_key(test_provider, test_key)
        )
        self.assertTrue(result)
        
        # Retrieve the key
        retrieved_key = self.loop.run_until_complete(
            self.keystore.get_key(test_provider)
        )
        self.assertEqual(retrieved_key, test_key)
        
        # Clean up
        self.loop.run_until_complete(
            self.keystore.delete_key(test_provider)
        )
    
    def test_key_info(self):
        """Test getting key info without exposing the key"""
        test_provider = "info_provider"
        test_key = f"info_api_key_{int(datetime.now().timestamp())}"
        
        # Store the key
        self.loop.run_until_complete(
            self.keystore.set_key(test_provider, test_key, {
                "name": "test info key"
            })
        )
        
        # Get key info
        key_info = self.loop.run_until_complete(
            self.keystore.get_key_info(test_provider)
        )
        
        # Verify info
        self.assertIsNotNone(key_info)
        self.assertEqual(key_info["provider"], test_provider)
        self.assertEqual(key_info["name"], "test info key")
        self.assertNotIn("key", key_info)
        
        # Clean up
        self.loop.run_until_complete(
            self.keystore.delete_key(test_provider)
        )
    
    def test_key_rotation(self):
        """Test rotating a key to a new value"""
        test_provider = "rotation_provider"
        original_key = f"original_key_{int(datetime.now().timestamp())}"
        rotated_key = f"rotated_key_{int(datetime.now().timestamp())}"
        
        # Store the key
        self.loop.run_until_complete(
            self.keystore.set_key(test_provider, original_key)
        )
        
        # Rotate the key
        result = self.loop.run_until_complete(
            self.keystore.rotate_key(test_provider, rotated_key)
        )
        self.assertTrue(result)
        
        # Verify the key was rotated
        retrieved_key = self.loop.run_until_complete(
            self.keystore.get_key(test_provider)
        )
        self.assertEqual(retrieved_key, rotated_key)
        
        # Verify key info includes rotation data
        key_info = self.loop.run_until_complete(
            self.keystore.get_key_info(test_provider)
        )
        self.assertIn("rotated_from", key_info)
        
        # Clean up
        self.loop.run_until_complete(
            self.keystore.delete_key(test_provider)
        )
    
    def test_expired_keys(self):
        """Test behavior with expired keys"""
        test_provider = "expired_provider"
        test_key = f"expired_key_{int(datetime.now().timestamp())}"
        
        # Set a key that's already expired
        yesterday = datetime.now() - timedelta(days=1)
        self.loop.run_until_complete(
            self.keystore.set_key(test_provider, test_key, {
                "expires_at": yesterday.isoformat()
            })
        )
        
        # Try to get the expired key
        retrieved_key = self.loop.run_until_complete(
            self.keystore.get_key(test_provider)
        )
        self.assertIsNone(retrieved_key)
        
        # Verify key info shows as expired
        key_info = self.loop.run_until_complete(
            self.keystore.get_key_info(test_provider)
        )
        self.assertTrue(key_info["is_expired"])
        
        # Clean up
        self.loop.run_until_complete(
            self.keystore.delete_key(test_provider)
        )
    
    def test_listing_providers(self):
        """Test listing all providers"""
        # Add several test keys
        providers = ["list_test_1", "list_test_2", "list_test_3"]
        for provider in providers:
            self.loop.run_until_complete(
                self.keystore.set_key(provider, f"key_for_{provider}")
            )
        
        # List all providers
        listed_providers = self.loop.run_until_complete(
            self.keystore.list_providers()
        )
        
        # Verify all test providers are in the list
        for provider in providers:
            self.assertIn(provider, listed_providers)
        
        # Clean up
        for provider in providers:
            self.loop.run_until_complete(
                self.keystore.delete_key(provider)
            )
    
    def test_persistence(self):
        """Test that keys persist across instances"""
        test_provider = "persistence_test"
        test_key = f"persistence_key_{int(datetime.now().timestamp())}"
        
        # Store a key
        self.loop.run_until_complete(
            self.keystore.set_key(test_provider, test_key)
        )
        
        # Create a new keystore instance
        new_keystore = Keystore(metadata={
            "encryption_key": self.test_key,
            "storage_location": self.test_dir
        })
        self.loop.run_until_complete(new_keystore.init())
        
        # Retrieve the key from the new instance
        retrieved_key = self.loop.run_until_complete(
            new_keystore.get_key(test_provider)
        )
        
        # Verify it's the same key
        self.assertEqual(retrieved_key, test_key)
        
        # Clean up
        self.loop.run_until_complete(
            self.keystore.delete_key(test_provider)
        )
    
    def test_self_test_method(self):
        """Test the module's own test method"""
        test_result = self.keystore.test()
        
        # Verify test structure
        self.assertIsInstance(test_result, dict)
        self.assertIn("success", test_result)
        self.assertIn("module", test_result)
        self.assertEqual(test_result["module"], "keystore")
        
        # Check security info
        self.assertIn("security", test_result)
        self.assertIn("cryptography_available", test_result["security"])

if __name__ == "__main__":
    # Run unit tests and output results in JSON format
    test_loader = unittest.TestLoader()
    test_suite = test_loader.loadTestsFromTestCase(TestKeystore)
    test_runner = unittest.TextTestRunner(verbosity=2)
    test_result = test_runner.run(test_suite)
    
    # Output results in JSON format
    results = {
        "total": test_result.testsRun,
        "failures": len(test_result.failures),
        "errors": len(test_result.errors),
        "success": test_result.wasSuccessful()
    }
    
    print(f"\nTest Summary: {json.dumps(results)}")

File: test/js/test_keystore.js
import { expect } from 'chai';
import path from 'path';
import fs from 'fs';
import os from 'os';
import { Keystore } from '../../hallucinate_app/node/keystore.js';

describe('Keystore Module', function() {
  // Test data
  const testMasterKey = 'test_master_key_do_not_use_in_production';
  const testStorageLocation = path.join(os.tmpdir(), 'hallucinate_app_test_keystore');
  const testProvider = 'test_provider';
  const testKey = 'test_api_key_' + Date.now();
  
  // Create a clean instance for testing
  let keystore;
  
  // Clean up test directory before tests
  before(function() {
    if (fs.existsSync(testStorageLocation)) {
      fs.rmSync(testStorageLocation, { recursive: true, force: true });
    }
  });
  
  it('should create a new keystore instance', async function() {
    keystore = new Keystore({
      encryptionKey: testMasterKey,
      storageLocation: testStorageLocation
    });
    
    expect(keystore).to.be.an.instanceOf(Keystore);
    expect(keystore.initialized).to.equal(false);
  });
  
  it('should initialize the keystore', async function() {
    const result = await keystore.init();
    expect(result).to.be.true;
    expect(keystore.initialized).to.be.true;
  });
  
  it('should create the storage directory if it does not exist', function() {
    expect(fs.existsSync(testStorageLocation)).to.be.true;
  });
  
  it('should store a key in the keystore', async function() {
    const result = await keystore.setKey(testProvider, testKey);
    expect(result).to.be.true;
  });
  
  it('should retrieve a key from the keystore', async function() {
    const key = await keystore.getKey(testProvider);
    expect(key).to.equal(testKey);
  });
  
  it('should get key info without exposing the key', async function() {
    const keyInfo = await keystore.getKeyInfo(testProvider);
    expect(keyInfo).to.be.an('object');
    expect(keyInfo.provider).to.equal(testProvider);
    expect(keyInfo).to.not.have.property('key');
  });
  
  it('should rotate a key', async function() {
    const newKey = 'rotated_key_' + Date.now();
    const result = await keystore.rotateKey(testProvider, newKey);
    expect(result).to.be.true;
    
    // Verify the key was rotated
    const key = await keystore.getKey(testProvider);
    expect(key).to.equal(newKey);
  });
  
  it('should list all providers with stored keys', async function() {
    const providers = await keystore.listProviders();
    expect(providers).to.be.an('array');
    expect(providers).to.include(testProvider);
  });
  
  it('should delete a key from the keystore', async function() {
    const result = await keystore.deleteKey(testProvider);
    expect(result).to.be.true;
    
    // Verify the key was deleted
    const key = await keystore.getKey(testProvider);
    expect(key).to.be.null;
  });
  
  it('should persist keys across instances', async function() {
    // Store a key in the first instance
    const testProvider2 = 'test_provider_2';
    const testKey2 = 'test_key_2_' + Date.now();
    await keystore.setKey(testProvider2, testKey2);
    
    // Create a new instance
    const keystore2 = new Keystore({
      encryptionKey: testMasterKey,
      storageLocation: testStorageLocation
    });
    await keystore2.init();
    
    // Verify the key is present
    const key = await keystore2.getKey(testProvider2);
    expect(key).to.equal(testKey2);
    
    // Clean up
    await keystore2.deleteKey(testProvider2);
  });
  
  it('should handle expired keys', async function() {
    // Store a key with expiration in the past
    const expiredProvider = 'expired_provider';
    const expiredKey = 'expired_key';
    
    const yesterday = new Date();
    yesterday.setDate(yesterday.getDate() - 1);
    
    await keystore.setKey(expiredProvider, expiredKey, {
      expiresAt: yesterday
    });
    
    // Try to get the expired key
    const key = await keystore.getKey(expiredProvider);
    expect(key).to.be.null;
    
    // Check key info
    const keyInfo = await keystore.getKeyInfo(expiredProvider);
    expect(keyInfo.isExpired).to.be.true;
    
    // Clean up
    await keystore.deleteKey(expiredProvider);
  });
  
  it('should run comprehensive self-tests', async function() {
    const testResults = await keystore.test();
    expect(testResults.success).to.be.true;
    expect(testResults.module).to.equal('keystore');
    expect(testResults.initialization).to.be.true;
    expect(testResults.persistence).to.be.true;
  });
  
  // Clean up after all tests
  after(function() {
    if (fs.existsSync(testStorageLocation)) {
      fs.rmSync(testStorageLocation, { recursive: true, force: true });
    }
  });
});

// Run test if this file is executed directly
if (import.meta.url === `file://${process.argv[1]}`) {
  // Create Mocha instance and run tests
  const Mocha = await import('mocha');
  const mocha = new Mocha.default();
  mocha.addFile(process.argv[1]);
  mocha.run(function(failures) {
    process.exit(failures ? 1 : 0);
  });
}